File: src/services/market-service.ts
//servies/market-service.ts

import pool from '../config/database';

export class MarketService {
    static async generateOneMinuteCandles() {
        const client = await pool.connect();
        try {
            console.log('⏳ Generating 1-minute candles...');

            // 1. Ambil semua saham aktif
            const stocks = await client.query('SELECT id, symbol FROM stocks WHERE is_active = true');

            for (const stock of stocks.rows) {
                try {
                    // 2. Ambil trades dalam 1 menit terakhir (Misal sekarang 10:05, ambil data 10:04:00 - 10:04:59)
                    const tradesRes = await client.query(`
                        /* dialect: postgres */
                        SELECT price, quantity, created_at 
                        FROM trades 
                        WHERE sell_order_id IN (SELECT id FROM orders WHERE stock_id = $1)
                        AND created_at >= date_trunc('minute', NOW()) - INTERVAL '1 minute'
                        AND created_at < date_trunc('minute', NOW())
                        ORDER BY created_at ASC
                    `, [stock.id]);

                    const trades = tradesRes.rows;
                    let open, high, low, close, volume;
                    const startTime = new Date();
                    startTime.setMinutes(startTime.getMinutes() - 1);
                    startTime.setSeconds(0);
                    startTime.setMilliseconds(0);

                    if (trades.length > 0) {
                        // KASUS A: Ada Transaksi
                        open = parseFloat(trades[0].price);
                        close = parseFloat(trades[trades.length - 1].price);

                        // Cari High & Low pakai Math.max/min
                        const prices = trades.map((t: any) => parseFloat(t.price));
                        high = Math.max(...prices);
                        low = Math.min(...prices);

                        // Hitung total volume
                        volume = trades.reduce((sum: number, t: any) => sum + t.quantity, 0);
                    } else {
                        // KASUS B: Tidak Ada Transaksi (Gunakan Close candle sebelumnya)
                        const lastCandle = await client.query(`
                            /* dialect: postgres */
                            SELECT close_price FROM stock_candles 
                            WHERE stock_id = $1 
                            ORDER BY start_time DESC LIMIT 1
                        `, [stock.id]);

                        // Null-safe guard: ensure lastCandle is not null/undefined before accessing rowCount
                        if ((lastCandle?.rowCount ?? 0) > 0) {
                            const lastPrice = parseFloat(lastCandle.rows[0].close_price);
                            open = lastPrice;
                            high = lastPrice;
                            low = lastPrice;
                            close = lastPrice;
                            volume = 0;
                        } else {
                            continue; // Belum ada data sama sekali, skip
                        }
                    }

                    // 3. Simpan ke stock_candles (untuk 1-minute raw data)
                    await client.query(`
                        /* dialect: postgres */
                        INSERT INTO stock_candles (stock_id, resolution, open_price, high_price, low_price, close_price, volume, start_time)
                        VALUES ($1, '1M', $2, $3, $4, $5, $6, $7)
                    `, [stock.id, open, high, low, close, volume, startTime]);

                    // 4. Juga simpan ke candles table untuk multi-timeframe support (jika table ada)
                    try {
                        await client.query(`
                            /* dialect: postgres */
                            INSERT INTO candles (stock_id, timeframe, open_price, high_price, low_price, close_price, volume, timestamp)
                            VALUES ($1, '1m', $2, $3, $4, $5, $6, $7)
                            ON CONFLICT (stock_id, timeframe, timestamp) DO UPDATE
                            SET open_price = EXCLUDED.open_price,
                                high_price = EXCLUDED.high_price,
                                low_price = EXCLUDED.low_price,
                                close_price = EXCLUDED.close_price,
                                volume = EXCLUDED.volume
                        `, [stock.id, open, high, low, close, volume, startTime]);
                    } catch (err: any) {
                        // Skip jika table candles belum ada
                        if (err.code !== '42P01') throw err;
                    }

                    console.log(`🕯️ Candle generated for ${stock.symbol}: O:${open} H:${high} L:${low} C:${close} V:${volume}`);
                } catch (stockErr) {
                    console.error(`❌ Error processing stock ${stock.symbol}:`, stockErr);
                    // Continue to next stock
                }
            }

            // 5. Generate candles untuk timeframe lain (5m, 15m, 1h, 1d)
            await this.aggregateCandles(client);
        } catch (err) {
            console.error('❌ Error generating candles:', err);
        } finally {
            // CRITICAL: Always release the client
            client.release();
        }
    }

    // Aggregate candles dari 1m ke timeframe yang lebih besar
    private static async aggregateCandles(client: any) {
        try {
            const timeframes = [
                { name: '5m', minutes: 5 },
                { name: '15m', minutes: 15 },
                { name: '1h', minutes: 60 },
                { name: '1d', minutes: 1440 }
            ];

            const stocks = await client.query('SELECT id FROM stocks WHERE is_active = true');

            for (const stock of stocks.rows) {
                for (const tf of timeframes) {
                    // Ambil candles 1m dalam range waktu tertentu
                    const now = new Date();
                    const interval = tf.minutes;
                    const startOfPeriod = new Date(now);
                    startOfPeriod.setMinutes(Math.floor(now.getMinutes() / interval) * interval - interval, 0, 0);
                    const endOfPeriod = new Date(startOfPeriod);
                    endOfPeriod.setMinutes(startOfPeriod.getMinutes() + interval);

                    const candlesRes = await client.query(`
                        /* dialect: postgres */
                        SELECT 
                            (array_agg(open_price ORDER BY timestamp ASC))[1] as open,
                            MAX(high_price) as high,
                            MIN(low_price) as low,
                            (array_agg(close_price ORDER BY timestamp DESC))[1] as close,
                            SUM(volume) as volume
                        FROM candles
                        WHERE stock_id = $1 
                        AND timeframe = '1m'
                        AND timestamp >= $2 
                        AND timestamp < $3
                    `, [stock.id, startOfPeriod, endOfPeriod]);

                    const candle = candlesRes.rows[0];
                    if (candle && candle.open) {
                        await client.query(`
                            /* dialect: postgres */
                            INSERT INTO candles (stock_id, timeframe, open_price, high_price, low_price, close_price, volume, timestamp)
                            VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
                            ON CONFLICT (stock_id, timeframe, timestamp) DO UPDATE
                            SET open_price = EXCLUDED.open_price,
                                high_price = EXCLUDED.high_price,
                                low_price = EXCLUDED.low_price,
                                close_price = EXCLUDED.close_price,
                                volume = EXCLUDED.volume
                        `, [stock.id, tf.name, candle.open, candle.high, candle.low, candle.close, candle.volume || 0, startOfPeriod]);
                    }
                }
            }
        } catch (err: any) {
            // Skip jika table candles belum ada
            if (err.code !== '42P01') {
                console.error('Error aggregating candles:', err.message);
            }
        }
    }

    // Fungsi buat API ambil data grafik dari stock_candles (backward compatible)
    static async getCandles(symbol: string, timeframe: string = '1m', limit: number = 1000) {
        try {
            // Coba query dari table candles (multi-timeframe support)
            const result = await pool.query(`
                /* dialect: postgres */
                SELECT 
                    extract(epoch from timestamp) * 1000 as time,
                    open_price as open, 
                    high_price as high, 
                    low_price as low, 
                    close_price as close,
                    volume
                FROM candles c
                JOIN stocks s ON c.stock_id = s.id
                WHERE s.symbol = $1
                AND c.timeframe = $2
                ORDER BY timestamp ASC
                LIMIT $3
            `, [symbol, timeframe, limit]);

            return result.rows;
        } catch (err: any) {
            // Fallback ke stock_candles jika table candles belum ada
            if (err.code === '42P01') { // Table doesn't exist
                console.warn('⚠️ Table candles tidak ditemukan, fallback ke stock_candles');
                const result = await pool.query(`
                    /* dialect: postgres */
                    SELECT 
                        extract(epoch from start_time) * 1000 as time,
                        open_price as open, 
                        high_price as high, 
                        low_price as low, 
                        close_price as close,
                        volume
                    FROM stock_candles sc
                    JOIN stocks s ON sc.stock_id = s.id
                    WHERE s.symbol = $1
                    ORDER BY start_time ASC
                    LIMIT $2
                `, [symbol, limit]);

                return result.rows;
            }
            throw err;
        }
    }

    // Fungsi untuk mengambil daily stock data (OHLC per session)
    static async getDailyStockData(symbol?: string) {
        let query = `
            /* dialect: postgres */
            SELECT 
                s.symbol,
                s.name,
                ts.session_number,
                ts.status as session_status,
                ts.started_at,
                ts.ended_at,
                d.prev_close,
                d.open_price,
                d.high_price,
                d.low_price,
                d.close_price,
                d.ara_limit,
                d.arb_limit,
                d.volume
            FROM daily_stock_data d
            JOIN stocks s ON d.stock_id = s.id
            JOIN trading_sessions ts ON d.session_id = ts.id
        `;

        const params: any[] = [];
        if (symbol) {
            query += ' WHERE s.symbol = $1';
            params.push(symbol);
        }

        query += ' ORDER BY ts.session_number DESC, s.symbol ASC LIMIT 100';

        const result = await pool.query(query, params);
        return result.rows;
    }
}

File: go-backend/services/market_service.go
package services

import (
	"context"
	"log"
	"time"

	"mbit-backend-go/config"

	"github.com/jackc/pgx/v5"
)

type MarketService struct{}

func (s *MarketService) GenerateOneMinuteCandles() error {
	ctx := context.Background()

	// Get active session
	var sessionId int
	err := config.DB.QueryRow(ctx, "SELECT id FROM trading_sessions WHERE status = 'OPEN' ORDER BY id DESC LIMIT 1").Scan(&sessionId)
	if err == pgx.ErrNoRows {
		// No open session, skip
		return nil
	} else if err != nil {
		return err
	}

	// Logic from src/services/market-service.ts
	// Aggregate trades from the last minute (or specific time window)
	// Node implementation usually truncates time to minute.

	now := time.Now()
	// Format to start of current minute (or previous minute?)
	// Cron runs at minute start. Usually we generate for the *previous* minute.
	// e.g. at 10:01:00 we generate for 10:00:00 - 10:00:59.

	prevMinute := now.Add(-1 * time.Minute).Truncate(time.Minute)

	log.Printf("🕯️ Generating candles for %s...", prevMinute)

	// 1. Get stocks with trades in that minute
	// We can do this in one big SQL query.

	// Note: Schema `stock_candles` index might be unique on (stock_id, resolution, start_time) ??
	// Check schema: `create unique index candles_stock_id_timeframe_timestamp_key on candles ...`
	// Wait, table is `candles` or `stock_candles`?
	// Schema has `candles` AND `stock_candles`.
	// Node `MarketService` uses which one?
	// `tes-canle.http` implies `candles`.
	// `db/ALL_schema_database.sql` has both.
	// `candles` table has `timeframe` column. `stock_candles` has `resolution`.
	// Node cron says `MarketService.generateOneMinuteCandles()`.
	// I'll check `db/migration_add_candles_watchlist.sql` or similar if I could.
	// But `candles` seems to be the one with `timeframe` varchar(10).
	// Let's assume `candles` table is the modern one.

	// Re-writing query for `candles` table
	queryCandles := `
		INSERT INTO candles (stock_id, session_id, timeframe, timestamp, open_price, high_price, low_price, close_price, volume)
		SELECT
			t.stock_id,
			$1,
			'1M',
			$2,
			(SELECT price FROM trades t2 WHERE t2.stock_id = t.stock_id AND t2.executed_at >= $2 AND t2.executed_at < $2 + interval '1 minute' ORDER BY t2.executed_at ASC LIMIT 1),
			MAX(t.price),
			MIN(t.price),
			(SELECT price FROM trades t3 WHERE t3.stock_id = t.stock_id AND t3.executed_at >= $2 AND t3.executed_at < $2 + interval '1 minute' ORDER BY t3.executed_at DESC LIMIT 1),
			SUM(t.quantity)
		FROM trades t
		WHERE t.executed_at >= $2 AND t.executed_at < $2 + interval '1 minute'
		GROUP BY t.stock_id
		ON CONFLICT (stock_id, timeframe, timestamp) DO NOTHING
	`

	tag, err := config.DB.Exec(ctx, queryCandles, sessionId, prevMinute)
	if err != nil {
		log.Printf("❌ Candle generation failed: %v", err)
		return err
	}

	log.Printf("✅ Generated %d candles", tag.RowsAffected())
	return nil
}

var GlobalMarketService = &MarketService{}
